core/src/main/java/org/apache/calcite/rel/rules/MinusToDistinctRule.java

```diff
@@ -0,0 +1,179 @@
+ *   except
+ *  select a,b from t3
+ * }</pre></blockquote>
+ *
+ * <p> will convert to
+ *
+ * <blockquote><pre>{@code
+ *  select a,b
+ *  from (select a,b,0 as m from t1
+ *  union all
+ *  select a,b,1 as m from t2
+ *  union all
+ *  select a,b,2 as m from t3)
+ *  group by a,b
+ *  having count(*) filter (where m = 0) > 0
+ *   and count(*) filter (where m = 1) = 0
+ *   and count(*) filter (where m = 2) = 0
+ * }</pre></blockquote>
+ *
+ * @see CoreRules#MINUS_TO_DISTINCT
+ */
+@Value.Enclosing
+public class MinusToDistinctRule
+    extends RelRule<MinusToDistinctRule.Config>
+    implements TransformationRule {
+
+  protected MinusToDistinctRule(Config config) {
+    super(config);
+  }
+
+  @Deprecated // to be removed before 2.0
+  public MinusToDistinctRule(Class<? extends Minus> minusClass,
+      RelBuilderFactory relBuilderFactory) {
+    this(MinusToDistinctRule.Config.DEFAULT.withRelBuilderFactory(relBuilderFactory)
+        .as(MinusToDistinctRule.Config.class)
+        .withOperandFor(minusClass));
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Minus minus = call.rel(0);
+
+    if (minus.all) {
+      // Nothing we can do
+      return;
+    }
+
+    final RelOptCluster cluster = minus.getCluster();
+    final RelBuilder relBuilder = call.builder();
+    final RexBuilder rexBuilder = cluster.getRexBuilder();
+    final int branchCount = minus.getInputs().size();
+
+    // For each child branch in minus,add a column which indicates branch index
+    // e.g. select EMPNO from emp -> select EMPNO,0 from emp
+    // 0 indicates that it comes from minus the first child branch
+    for (int i = 0; i < branchCount; i++) {
+      relBuilder.push(minus.getInput(i));
+      relBuilder.projectPlus(relBuilder.literal(new BigDecimal(i)));
+    }
+
+    // create a union above all the branches
+    relBuilder.union(true, branchCount);
+
+    final RelNode union = relBuilder.peek();
+    final int originalFieldCnt = union.getRowType().getFieldCount() - 1;
+
+    ImmutableList.Builder<RexNode> projects = ImmutableList.builder();
+    // skip the branch index column
+    projects.addAll(Util.first(relBuilder.fields(), originalFieldCnt));
+
+    // On top of the Union, add a Project and add branch cnt number boolean columns
+    // e.g. LogicalProject(EMPNO=[$0], $f1=[=($1, 0)], $f2=[=($1, 1)], $f3=[=($1, 2)])
+    // $f1,$f2,$f3 are the boolean indicate whether it comes from the corresponding branch
+    for (int i = 0; i < branchCount; i++) {
+      projects.add(
+          relBuilder.equals(relBuilder.field(originalFieldCnt),
+              relBuilder.literal(new BigDecimal(i))));
+    }
+
+    relBuilder.project(projects.build());
+
+    // Add the count(*) filter $f1(..) for each branch
+    ImmutableList.Builder<RelBuilder.AggCall> aggCalls = ImmutableList.builder();
+    for (int i = 0; i < branchCount; i++) {
+      aggCalls.add(relBuilder.countStar(null).filter(relBuilder.field(originalFieldCnt + i)));
+    }
```

Comment:
Is there anyway to combine these `for` loops? I noticed we iterate through the branch count multiple times.

Reply:
Yes,adding the Count Filter is mainly to calculate the specific value of each branch with group by fields.
![image](https://github.com/apache/calcite/assets/8057451/8210b672-1fca-4d6e-80c1-c151446c9536)